File: src/main/java/Benchmarks/Prog2/Prog2.py
import time
start_time = time.time()
def contain(s , End):
    for x in range(len(s)):
        if(s[x] == End):
            return True
    return False

def findPath(Start , End):
    s = []
    s.append(Start)
    while(contain(s , End) == False):
        if(len(s) > 30000):
            return
        r = s.pop(0)
        s += checkRules(r)

def checkRules(s):
    a = []
    if s[-1] == 'I':
        a.append(s + 'U')
    a.append(s + s[1:])
    for i in range(len(s)- 2) :
        if s[i] == 'I' and s[i+1] == 'I' and s[i+2] == 'I':
            a.append(s[:i] + 'U' +s[i+3:])
    for i in range(0,len(s)-1):
        if s[i] == 'U' and s[i + 1] == 'U':
            a.append(s[:i] + s[i+2:])
    return a

findPath("MI", "MUIU")
findPath("MI" , "MIUIUIUIU")
findPath("MI" , "MUIIU")
findPath("MI" , "MIUIIIIUIUIIIIU")
findPath("MI" , "MUIU")
findPath("MI" , "MIUIUIUIU")
findPath("MI" , "MUIIU")
findPath("MI" , "MIIUUII")
findPath("MI" , "Notpos")
findPath("MI" , "Notpos")
print((time.time() - start_time) * 1000)

File: src/main/java/Benchmarks/Prog2/Prog2.java
package Benchmarks.Prog2;

import java.util.ArrayList;

public class Prog2 {

    public static void main(String[] args) {
        long start = System.currentTimeMillis() ;
        findPath("MI", "MUIU");
        findPath("MI" , "MIUIUIUIU");
        findPath("MI" , "MUIIU");
        findPath("MI" , "MIUIIIIUIUIIIIU");
        findPath("MI" , "MUIU");
        findPath("MI" , "MIUIUIUIU");
        findPath("MI" , "MUIIU");
        findPath("MI" , "MIIUUII");
        findPath("MI" , "Notpos");
        findPath("MI" , "Notpos");
        long end = System.currentTimeMillis() ;
        long diff = end - start ;
        System.out.println(diff);
    }

    private static void findPath(String start , String End) {
        ArrayList<String> paths = new ArrayList<>();
        paths.add(start);
        while(!paths.contains(End)){
            if(paths.size() > 30000){
                return;
            }
            String s = paths.get(0);
            paths.remove(0);
            ArrayList<String> newPaths = checkRules(s);
            paths.addAll(newPaths);
        }
    }

    private static ArrayList<String> checkRules(String start) {
        ArrayList<String> newPaths = new ArrayList<>();
        if(start.endsWith("I")){
            newPaths.add(start + "U");
        }
        newPaths.add( start + start.substring(1));
        for(int i = 0 ; i < start.length() - 3;i++){
            if(start.charAt(i) == 'I' && start.charAt(i + 1) == 'I' && start.charAt(i + 2) == 'I'){
                newPaths.add(start.substring(0 , i) + "U" + start.substring(i +3));
            }
        }
        for(int i = 0 ; i < start.length() - 2;i++){
            if(start.charAt(i) == 'U' && start.charAt(i + 1) == 'U' ){
                newPaths.add(start.substring(0 , i) + start.substring(i + 2));
            }
        }
        return newPaths ;
    }

}
